### ISSUE TITLE
Python bindings: Call for help from the community

### ISSUE THREAD

**Issue by jmarrec** [CoolProp/CoolProp#2119]
A recent effort enabled building the shared libraries and python bindings directly on Github Actions. Support for Python 3.9 and 3.10 was added. 

**We need your help to test these out. This will only take a few minutes of your time, yet it is absolutely critical for us to be aware of bugs.**

The only thing you need to do is to install from testpypi and test them.

```shell
pip install -i https://test.pypi.org/simple/ CoolProp==6.4.2.dev0

# Short test
python -c 'from CoolProp.CoolProp import get_global_param_string; print("CoolProp gitrevision:", get_global_param_string("gitrevision"))'
```

**You can then use the following template to report what you tested and whether it worked or not**

```
*  Operating System+version(+arch if not x86_64/amd64): [eg: **Linux Ubuntu 18.04**]
* `python --version`: [eg: Python 3.9.7]  
* [ ] The bindings work (if they don't, please include information to reproduce / stack trace)
```

---

## Summary of tests:

 (edit in place)

|          | Windows 10 | Ubuntu | macOS Intel | macOS arm64 |
|--------|------------------|------------|------------------|---------------------|
| 3.7   |    #2120        |  OK       |         OK        |         N/A          |
| 3.8   |        OK         |  OK       |         OK        |                         |
| 3.9   |        OK         |  OK       |         OK        |                         |
| 3.10 |        OK         |  OK       |         OK        |                         |

**Comment by jmarrec** [comment]
*  Operating System+version(+arch if not x86_64/amd64): Windows 10 x86_64
* `python --version`:  Python 3.10.1
* [x] The bindings work (if they don't, please include information to reproduce / stack trace)

---

*  Operating System+version(+arch if not x86_64/amd64): Windows 10 x86_64
* `python --version`:  Python 3.9.5
* [x] The bindings work (if they don't, please include information to reproduce / stack trace)

----

*  Operating System+version(+arch if not x86_64/amd64): Windows 10 x86_64
* `python --version`:  Python 3.8.5
* [x] The bindings work (if they don't, please include information to reproduce / stack trace)

**Comment by jmcarter17** [comment]
- macOS 12.3.1 (Intel) x86_64
- Python 3.10.3
- [x] The bindings work

**Comment by jmcarter17** [comment]
*  Operating System+version(+arch if not x86_64/amd64): macOS 12.3.1 (Intel)
* `python --version`: Python 3.8.13
* [x] The bindings work (if they don't, please include information to reproduce / stack trace)

**Comment by jmcarter17** [comment]
*  Operating System+version(+arch if not x86_64/amd64): macOS 12.3.1 (Intel)
* `python --version`: Python 3.9.11
* [x] The bindings work (if they don't, please include information to reproduce / stack trace)

**Comment by jmcarter17** [comment]
*  Operating System+version(+arch if not x86_64/amd64): macOS 12.3.1 (Intel)
* `python --version`: Python 3.7.13
* [x] The bindings work (if they don't, please include information to reproduce / stack trace)

**Comment by jmarrec** [comment]
@jmcarter17 Awesome, thanks a bunch!

**Comment by bonnema** [comment]
Any plans for Windows 32-bit?

**Comment by jmarrec** [comment]
I personally invested (**waaayyyy**) too much time on this project I don't even really use :)

Happy to have anyone willing take a crack at fixing the windows stuff. Apparently it's having an issue with MSBuild.exe, and is picking up the adm64 version: https://github.com/CoolProp/CoolProp/runs/5981837331?check_suite_focus=true

```
   -- COOLPROP_INSTALL_PREFIX=D:/a/CoolProp/CoolProp/install_root
    CMake Error at CMakeLists.txt:146 (project):
      Failed to run MSBuild command:
  
        C:/Program Files/Microsoft Visual Studio/2022/Enterprise/MSBuild/Current/Bin/amd64/MSBuild.exe
  
      to get the value of VCTargetsPath:
  
        Microsoft (R) Build Engine version 17.1.0+ae57d105c for .NET Framework
        Copyright (C) Microsoft Corporation. All rights reserved.
  
        Build started 4/12/2022 7:44:29 AM.
        Project "D:\a\CoolProp\CoolProp\wrappers\Python\cmake_build\vc17-32bit\CMakeFiles\3.23.0\VCTargetsPath.vcxproj" on node 1 (default targets).
        C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(437,5): error MSB8013: This project doesn't contain the Configuration and Platform combination of Debug|Win32. [D:\a\CoolProp\CoolProp\wrappers\Python\cmake_build\vc17-32bit\CMakeFiles\3.23.0\VCTargetsPath.vcxproj]
        Done Building Project "D:\a\CoolProp\CoolProp\wrappers\Python\cmake_build\vc17-32bit\CMakeFiles\3.23.0\VCTargetsPath.vcxproj" (default targets) -- FAILED.
  
        Build FAILED.
  
        "D:\a\CoolProp\CoolProp\wrappers\Python\cmake_build\vc17-32bit\CMakeFiles\3.23.0\VCTargetsPath.vcxproj" (default target) (1) ->
        (PrepareForBuild target) ->
          C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(437,5): error MSB8013: This project doesn't contain the Configuration and Platform combination of Debug|Win32. [D:\a\CoolProp\CoolProp\wrappers\Python\cmake_build\vc17-32bit\CMakeFiles\3.23.0\VCTargetsPath.vcxproj]
  
            0 Warning(s)
            1 Error(s)
  
        Time Elapsed 00:00:15.71
  
  
      Exit code: 1
```

I just tried with msvc 2019 and it's the same issue:
https://github.com/CoolProp/CoolProp/runs/5993732487?check_suite_focus=true

Someone could try reproducing manually:
```
vcvarsall.bat x86
COOLPROP_CMAKE=vc17,32
python setup.py bdist_wheel
```

I'm trying one last thing in https://github.com/CoolProp/CoolProp/pull/2120/commits/4f866e980b17da543bb41dcae72ed418b15bd2c6 (not using COOLPROP_CMAKE), then I give up and move on.

**Comment by jmarrec** [comment]
Seems like it worked in https://github.com/CoolProp/CoolProp/runs/5994039732?check_suite_focus=true
@ibell pointing out we may not actually need to use COOLPROP_CMAKE and could simplify the workflow file greatly to be something like https://github.com/ibell/pdsim/blob/master/.github/workflows/cibuildswheels.yml, but I don't have time to try it out (it gets really old really quickly to launch Github Actions run, wait 5minutes to see them fail, rinse and repeat)

**Comment by jmarrec** [comment]
To test the win32 (py38 py39): `pip install -i https://test.pypi.org/simple/ CoolProp==6.4.2.post1`

**Comment by dflorez0** [comment]
*  Operating System+version(+arch if not x86_64/amd64): Windows10 64bit
* `python --version`: Python 3.10.2
* [ X] The bindings work

**Comment by jmarrec** [comment]
@jmcarter17 for arm64, I identified a fix in my new PR that aims to simplfy the cibuildwheel setting (and remove the use of COOLPROP_CMAKE) in #2122 and move it by itself into a minimal fix pr at #2123 (runners are still churning and won't post a new version to testpypi  until it's merged but artifacts will be available for testing via manual download in a few)

**Comment by ibell** [comment]
@jmarrec is there a reason no not build universal binary wheels on mac?

**Comment by jmarrec** [comment]
@ibell if we build both x86_64 and arm64, the universal2 is useless, because pip will always select the most precise available whl, so if you're on x86_64 it takes the x86_64, if you're on arm64 it'll take the arm64 one. The universal2 won't be picked up.
universal2 are also larger.

**Comment by flaviobn** [comment]
*  Operating System+version(+arch if not x86_64/amd64): Alpine Linux v3.15 (in the official python:3-alpine container)
* `python --version`: Python 3.10.4 
* [ ] The bindings work (if they don't, please include information to reproduce / stack trace)

This image or other similar images based on Alpine are often used to containerize Python applications. Alpine uses **musl** libc instead of **glibc**. CoolProp (and Python bindings) compiles from source and works properly on Alpine.

```
/ # pip install -i https://test.pypi.org/simple/ CoolProp==6.4.2.dev0
Looking in indexes: https://test.pypi.org/simple/
ERROR: Could not find a version that satisfies the requirement CoolProp==6.4.2.dev0 (from versions: none)
ERROR: No matching distribution found for CoolProp==6.4.2.dev0
```
Running pip again with -vvv and `pip debug --verbose`, I found out that Alpine only supports "musllinux" or "linux" tagged wheels. The "manylinux" wheels are not supported.

I don't know if the CoolProp team wants to support this platform. Maybe [Alpine Wheels](https://github.com/alpine-wheels) is a better place.

**Comment by ibell** [comment]
@flaviobn With https://github.com/CoolProp/CoolProp/commit/6f4dfa82c1232f4827a81040284262a7099ed3fe I added musl support.  Please try to install now

**Comment by flaviobn** [comment]
*  Operating System+version(+arch if not x86_64/amd64): Alpine Linux v3.15 (in the official python:3-alpine container)
* `python --version`: Python 3.10.4
* [X] The bindings work

**Comment by thomasboziuk** [comment]
*  Operating System+version(+arch if not x86_64/amd64): Linux Ubuntu 22.04
* `python --version`: Python 3.10.4
* [ x] The bindings work (if they don't, please include information to reproduce / stack trace)

the .dev0 install seems to be all good on the new ubuntu LTS

**Comment by fwitte** [comment]
*  Operating System+version(+arch if not x86_64/amd64): **Linux Ubuntu 20.04**
* `python --version`: Python 3.9.5  
* [x] The bindings work

**Comment by smyng91** [comment]
*  Operating System: **macOS 12.3.1 (arm64)**
* `python --version`: Python 3.9.12
* [x] The bindings work

**Comment by ibell** [comment]
It's not that simple sadly, because we need to get all the other wrappers (and the docs) also updated, and nobody has stepped up to shepherd that.

**Comment by bmarjaba** [comment]
Happy to see that CoolProp finally worked with arm64 in MATLAB! I am still unable to make it work though. I tried it with python 3.8, and 3.9.12 (same version as @smyng91 ). Sam, can you please help me get it to work? the error i am getting is when calling the command:
[v,e] = pyversion; system([e,' -m pip install --user -U CoolProp'])

**Comment by ibell** [comment]
Please try the install command from the top of this thread

**Comment by bmarjaba** [comment]
i ran <<pip install -i https://test.pypi.org/simple/ CoolProp==6.4.2.dev0>> in terminal, and it said requirement already satisfied. then i ran <<python -c 'from CoolProp.CoolProp import get_global_param_string; print("CoolProp gitrevision:", get_global_param_string("gitrevision"))'>> and i got errors. where am i supposed to run this last command? terminal or matlab's terminal? and do i run it as is? the way i wrote it?

**Comment by jmarrec** [comment]
Weird that you got requirements already satisfied. do `python -m pip uninstall CoolProp` then redo `python -m pip install -i https://test.pypi.org/simple/ CoolProp==6.4.2.dev0` and try running the `python -c` in the same terminal and report back.
Would be nice to know which OS/Platform you're using (eg: windows 10) and the python version you have too (python --version)

**Comment by bmarjaba** [comment]
I'm using macOS Monterey v12.3.1, and I have python 3.8 installed.

I ran the following in terminal:

marjaba@Bernards-MacBook-Pro-2 ~ % python3 -m pip uninstall CoolProp                                
WARNING: Skipping CoolProp as it is not installed.
marjaba@Bernards-MacBook-Pro-2 ~ % pip install -i https://test.pypi.org/simple/ CoolProp==6.4.2.dev0
Looking in indexes: https://test.pypi.org/simple/
Requirement already satisfied: CoolProp==6.4.2.dev0 in /Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages (6.4.2.dev0)
marjaba@Bernards-MacBook-Pro-2 ~ % python3 -c 'from CoolProp.CoolProp import get_global_param_string; print("CoolProp gitrevision:", get_global_param_string("gitrevision"))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'CoolProp'

**Comment by bmarjaba** [comment]
marjaba@Bernards-MacBook-Pro-2 ~ % which pip
/Library/Frameworks/Python.framework/Versions/3.8/bin/pip
marjaba@Bernards-MacBook-Pro-2 ~ % which python3
/Library/Frameworks/Python.framework/Versions/3.9/bin/python3

how do i align them?

**Comment by ibell** [comment]
Use python -m pip

This will ensure you get the pip linked with your python.

**Comment by bmarjaba** [comment]
do i have to do anything after that? because a list of command options were shown. then i did which pip and which python 3, i got the same result (one 3.8 and the other 3.9)

**Comment by ibell** [comment]
Put the normal install command after python -m pip .  The -m pip is only
used to ensure that you pick up the pip associated with the given instance
of python.  It's confusing,  I know...

**Comment by bmarjaba** [comment]
how do i select it? i don't really know my way around mac's terminal so sorry for the inconvenience.

- just to recap: i need to align pip with python, this is done by running _**python3 -m pip**_
- then i select the normal install command which is this: install Install packages? it's the very first option in the list?
how do i select that command option?

**Comment by ibell** [comment]
Do:  python -m pip install -i https://test.pypi.org/simple/ CoolProp

**Comment by bmarjaba** [comment]
successfully installed. now when i run the python3 -c command i got this:
marjaba@Bernards-MacBook-Pro-2 ~ % python3 -c 'from CoolProp.CoolProp import get_global_param_string; print("CoolProp gitrevision:", get_global_param_string("gitrevision"))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/CoolProp/__init__.py", line 16, in <module>
    from .CoolProp import AbstractState
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/CoolProp/CoolProp.cpython-39-darwin.so, 0x0002): symbol not found in flat namespace '__Z5PropsPKccdcdS0_'

**Comment by jmarrec** [comment]
> Weird that you got requirements already satisfied. do `python -m pip uninstall CoolProp` then redo `python -m pip install -i https://test.pypi.org/simple/ CoolProp==6.4.2.dev0` and try running the `python -c` command in the same terminal and report back.
> Would be nice to know which OS/Platform you're using (eg: windows 10) and the python version you have too (python --version)

Tldr: the issue is on your end. Notice how I specially avoided calling pip directly instead wrote `python -m ..`? This is specifically for a reason. I don't know how you ended up with a pip exe being out of sync with your python version, but that's something we can't really help you with

**Comment by jmarrec** [comment]
Personally one of my machines is a Mac and use pyenv and virtualenv (like I do on Ubuntu too) and I've never had thus problem. Check your PATH environment variable?

**Comment by bmarjaba** [comment]
i ran [v,e] = pyversion; system([e,' -m pip install --user -U CoolProp']) on MATLAB now and it successfully installed. CoolProp finally worked. Thank you very much guys! I assume coolprop is now working with excel as well on arm64?

**Comment by defencedog** [comment]
Not wroking arm64 Android Termux
```
~ $ pip --version
pip 22.2.2 from /data/data/com.termux/files/usr/lib/python3.9/site-packages/pip (python 3.9)
~ $ pip install -i https://test.pypi.org/simple/ CoolProp==6.4.2.dev0
Looking in indexes: https://test.pypi.org/simple/
ERROR: Could not find a version that satisfies the requirement CoolProp==6.4.2.dev0 (from versions: none)
ERROR: No matching distribution found for CoolProp==6.4.2.dev0
```
Using tarball at sourceforge `CoolProp-6.4.2.dev0.tar.gz`

```
CoolProp/CoolProp.cpp:65595:54: error: no member named 'tp_print' in '_typeobject'
  __pyx_type_8CoolProp_8CoolProp_PyPhaseEnvelopeData.tp_print = 0;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
CoolProp/CoolProp.cpp:65603:53: error: no member named 'tp_print' in '_typeobject'
  __pyx_type_8CoolProp_8CoolProp_PyGuessesStructure.tp_print = 0;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
CoolProp/CoolProp.cpp:65611:50: error: no member named 'tp_print' in '_typeobject'
  __pyx_type_8CoolProp_8CoolProp_PyCriticalState.tp_print = 0;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
CoolProp/CoolProp.cpp:65619:49: error: no member named 'tp_print' in '_typeobject'
  __pyx_type_8CoolProp_8CoolProp_PySpinodalData.tp_print = 0;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
CoolProp/CoolProp.cpp:65789:48: error: no member named 'tp_print' in '_typeobject'
  __pyx_type_8CoolProp_8CoolProp_AbstractState.tp_print = 0;
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
CoolProp/CoolProp.cpp:65825:40: error: no member named 'tp_print' in '_typeobject'
  __pyx_type_8CoolProp_8CoolProp_State.tp_print = 0;
```

**Comment by jmarrec** [comment]
@defencedog android + termux + arm, that's quite a corner case and I wouldn't expect it to work.

@ibell Any plans to push this to pypi at last? Seems like there's been a bunch of testing in almost 6 months

**Comment by defencedog** [comment]
@jmarrec  it is working flawless & I have reported to author; for your future reference I also have a working [cantera](https://github.com/Cantera/cantera) install & use it regularly on jupyter notebook on Galaxy Tab S7

**Comment by jmarrec** [comment]
> @jmarrec  it is working flawless & I have reported to author; for your future reference I also have a working [cantera](https://github.com/Cantera/cantera) install & use it regularly on jupyter notebook on Galaxy Tab S7

 Didn't see were you said it worked. That's kinda cool to hear!

**Comment by jowr** [comment]
Do I understand correctly that I have to add our pypi token and then the GitHub action automatically publishes the binaries of a git tag?

**Comment by jowr** [comment]
@ibell Could you add my user to the CoolProp project on PyPi?

**Comment by ibell** [comment]
Invited to PYPI!

**Comment by jmarrec** [comment]
> Do I understand correctly that I have to add our pypi token and then the GitHub action automatically publishes the binaries of a git tag?

correct

https://github.com/CoolProp/CoolProp/blob/cb437eb7589b3384bd23ab4868da9bb30b7b25c7/.github/workflows/python_cibuildwheel.yml#L91-L97

**Comment by jowr** [comment]
Yes - I think we are cool now.

The only thing that needs fixing are the docs. They do not build properly and I guess that is a Python 2 vs 3 issue. Please have a look at the actions related to my latest branch that should fix the docs. To be continued.

**Comment by ibell** [comment]
I removed non-ascii characters in the bib file
in d818f86c7f1aec3f6cc23862481c21a2bb2c826c
